_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,493 @@
+You can imagine the scenario, where you will change some custom resource value of Kafka and then you will want to assert
+that this value was correctly updated. This is a very good example where `race condition` comes into play. This race condition
+can occur if the assertion (verify phase) is too quick and value has not been updated. That is the reason why we have this dynamic `waitFor()` function. The logic behind this method is that every `pollIntervalMs` for specific period of time `timeoutMs` the method checks
+`BooleanSupplier` condition which user can provide and if this condition is true the method will successfully end. Otherwise, we will get the
+`WaitException` with some reasonable description. The method implementation can be found [here](https://github.com/strimzi/strimzi-kafka-operator/blob/master/test/src/main/java/io/strimzi/test/TestUtils.java#L115-L145).
+
+#### Constants
+
+Purpose of this class is to keep global constants used across system tests in one place. You can find there mentioned constants like timeout interval and poll interval used by `waitFor()` method. Also, there are various other timeouts for cluster operator and clients, definitions of specific ports such as KafkaBridge metric, TopicOperator metric, Keycloak and so on. Lastly, there are all tag names,
+which are used in whole `system test` module, for instance `ACCEPTANCE`, `REGRESSION` and more.
+ 
+#### Environment
+
+System tests can be configured by several environment variables, which are loaded before test execution.
+
+Variables can be defined via environmental variables or a configuration file. This file can be located anywhere on the file system as long as a path is provided to this file.
+The path is defined by environment variable `ST_CONFIG_PATH`. If it is not defined, the default config file location `systemtest/config.json` will be used instead.
+Loading of system configuration has the following priority order:
+1. Environment variable
+2. Variable defined in configuration file
+3. Default value
+
+Here is the list of important environment variables:
+
+1. `DOCKER_ORG` = Specifies the organization/repo containing the image used in system tests (default value `strimzi`)
+2. `DOCKER_TAG` = Specifies the image tags used in system tests (default value `latest`)
+3. `SKIP_TEARDOWN` = Skip teardown phase - primarily used for debug purposes (default value `false`)
+4. `ST_KAFKA_VERSION` =  Specifies Kafka version used in images during the system tests (default value `2.6.0`)
+
+If you want to see all our environment variables you can see it [here](https://github.com/strimzi/strimzi-kafka-operator/blob/master/development-docs/TESTING.md#environment-variables)
+
+#### Our Kafka clients
+
+Another helpful module of the system tests are Kafka clients, which contains three usage types:
+
+1. For internal communication within Kubernetes - using test-client image
+2. For internal communication within Kubernetes - client-examples image
```

Comment:
Might be worth explaining why two kinds of internal clients. Are there cases whn to use one and when to use another? Or is one of them _deprecated_ and should not be used for new tests?

Reply:
Yes, internal clients which are using the test-client image are deprecated I will mention this one.